pyheaders/compiler.py

```diff
@@ -177,39 +178,119 @@ def _find_compile_commands(start_at: AnyStr = None) -> AnyStr:
 
     @staticmethod
     @lru_cache
-    def __get_compile_commands(filename: AnyStr) -> CompileCommands:
+    def __get_compile_commands(filename: AnyStr) -> Tuple[AnyStr, CompileCommands]:
         commands_filename = CommandsParser._find_compile_commands(os.path.dirname(filename))
         if commands_filename and os.path.isfile(commands_filename):
             with open(commands_filename) as commands_file:
-                return json.load(commands_file)
-        return []
+                return os.path.dirname(commands_filename), json.load(commands_file)
+        return os.getcwd(), []
+
+    @staticmethod
+    @lru_cache
+    def __fix_path(start_at: AnyStr, cmd_path: AnyStr, check: Callable[[AnyStr], bool]) -> AnyStr:
+        '''
+        Attempts to fix a broken path.
+
+        @param start_at The absolute path to use as a base current path.
+        @param cmd_path The broken path from the command.
+        @param check    A predicate that determines whether the path is valid (os.path.isdir or
+                        os.path.isfile)
+
+        @returns AnyStr A valid (and hopefully correct) absolute path to use.
+        '''
+        cmd_path = cmd_path.lstrip(os.path.sep)
+        for reverse_sep_idx in range(cmd_path.count(os.path.sep)):
+            cmd_path_fragment = cmd_path.split(os.path.sep, reverse_sep_idx)[-1]
+            if start_at.endswith(cmd_path_fragment):
+                return start_at
+            if check(fixed_path := os.path.join(start_at, cmd_path_fragment)):
+                return fixed_path
+        return start_at
+
+    @staticmethod
+    @lru_cache
+    def __get_path(start_at: AnyStr, cmd_path: AnyStr, check: Callable[[AnyStr], bool]) -> AnyStr:
+        '''
+        Returns an absolute path from a path in the compile commands.
+        If the path in the compile commands is relative, it is converted to be relative to
+        `start_at`.
+        If the path is a missing absolute path, `__get_path()` attempts to fix it by slowly
+        removing directories from the beginning of the command's path and using the reminder
+        as a relative path.
+
+        @param start_at The path to use as a base current path.
+        @param cmd_path The path from the command.
+        @param check    A predicate that determines whether the path is valid (os.path.isdir or
+                        os.path.isfile)
+
+        @returns AnyStr A valid (and hopefully correct) absolute path to use.
+        '''
+        start_at = os.path.abspath(start_at)
+        path = os.path.join(start_at, cmd_path)
+        if os.path.isabs(cmd_path) and not check(path):
```

Comment:
If `cmd_path` is absolute, doesn't it mean that `path` won't pass check? or it is for the instance when start_at is ""?

Reply:
If a path that was passed to `os.path.join()` is absolute all parts that come before it are discarded.
So essentially, if `os.path.isabs(cmd_path)` then you'll get that `os.path.join(start_at, cmd_path) == cmd_path`.